Fix image test [WIP]

damienmarchal opened the pull request:
______________________________________________________
<!--- Please leave this at the end of your message -->
This PR: 
- [ ] builds with SUCCESS for all platforms on the CI.
- [ ] does not generate new warnings.
- [ ] does not generate new unit test failures.
- [ ] does not generate new scene test failures.
- [ ] does not break API compatibility.
- [ ] is more than 1 week old (or has fast-merge label).

**Reviewers will merge only if all these checks are true.**

damienmarchal committed:
[SofaSphFluid] Move from module to applications/plugins and update the dependency

damienmarchal committed:
[SofaSphFluid] still moving from modules to applications/plugins

damienmarchal committed:
[SofaMiscCollision] Move the module into a plugin.

damienmarchal committed:
[SofaMeshCollision] Update to take into account the plugininzation of SofaMiscCollision

damienmarchal committed:
[SofaMeshCollision] Remove commented includes.

damienmarchal committed:
[SceneCreator] Update to have a dependency on plugin.

damienmarchal committed:
[sofaProjectExamples] remove commented lines or unused.

damienmarchal committed:
[SofaSphFluid] Update CMakeList files to make it a real plugin.

damienmarchal committed:
[typedef] update all the typedef to remove the one about SOfaMiscCollision

damienmarchal committed:
[SofaUserInteraction] remove commented code.

damienmarchal committed:
[SofaComponentAdvanced] Remove SofaSphFluid depency

damienmarchal committed:
[SofaComponentMisc] Remove the dependency to SofaMiscCollision.

damienmarchal committed:
[SofaCUDA] Update to properly handle the dependency with plugin/SofaSphFluid

damienmarchal committed:
[Compilant] Add the proper dependency to SofaMiscCollision

A better approach would be to only activate compilant part when the SofaMiscCollision
is on. This is something todo...by someone who use compliant.

damienmarchal committed:
[SofaCUDA] Fix CMakeList with missing .cu files.

damienmarchal committed:
Merge remote-tracking branch 'upstream/master' into pluginize_sofasphfluid

# Conflicts:
#	SofaAdvanced/config.h.in
#	applications/plugins/CMakeLists.txt
#	modules/SofaComponentAdvanced/CMakeLists.txt
#	modules/SofaComponentAdvanced/initComponentAdvanced.cpp

damienmarchal committed:
[Registration] Add dependency to SofaMiscCollision plugin.

damienmarchal committed:
[all] todo.

damienmarchal committed:
[SceneChecker] Improve the checker.

damienmarchal committed:
[SofaKernel] Add a new option in CMakeList to add or not the deprecated components.

damienmarchal committed:
[TODO] put example of three set of components.

damienmarchal committed:
[all] some stats about the inclusion of files while compiling.

damienmarchal committed:
[CMakeList.txt] Add two new config options to allow us to handle software evolution

Software evolution is hard an requires care and strategy.
To ease this we have splitted the set of sofa files in three sets:
- Minimal (The component of SofaNG)
- Full
- Deprecated

The two added options works in the following way:
When SOFA_BUILD_FULLSETCOMPONENTS is ON the components in full are added minimal
When SOFA_BUILD_DEPRECATEDCOMPONENTS is ON the components in the deprecated set are added to full set.

This organization is orthogonal to Plugins which are more specific to "context"
specific component (like haptic, implicitfield or whatever).

damienmarchal committed:
[SofaBoundaryCondition] Update the CMakeList to initialize the three set approach of lifetime management.

damienmarchal committed:
[SofaConstraint] Update the CMakeList to initialize the three set approach of lifetime management.

damienmarchal committed:
[SofaDenseSolver] Update the CMakeList to initialize the three set approach of lifetime management.

damienmarchal committed:
[SofaExporter] Update the CMakeList to initialize the three set approach of lifetime management.

damienmarchal committed:
[SofaGeneralAnimationLoop] Update the CMakeList to initialize the three set approach of lifetime management.

damienmarchal committed:
[SofaGeneralDeformable] Update the CMakeList to initialize the three set approach of lifetime management.

damienmarchal committed:
[SofaGeneralEngine] Update the CMakeList to initialize the three set approach of lifetime management.

damienmarchal committed:
[SofaGeneralExplicitOdeSolver] Update the CMakeList to initialize the three set approach of lifetime management.

damienmarchal committed:
[typedefs] remove typedefs.

damienmarchal committed:
[SofaGraphComponent] Put the SceneCheckAPI in its own file and improve the implementation.

bcarrez committed:
Add SOFA_BUILD_COMPONENTSET cmake option (core/full(default)/deprecated)

bcarrez committed:
[SofaGeneralImplicitOdeSolver] Update the CMakeList to initialize the three set approach of lifetime management.

bcarrez committed:
[SofaGeneralLinearSolver] Update the CMakeList to initialize the three set approach of lifetime management.

bcarrez committed:
[SofaGeneralMeshCollision] Update the CMakeList to initialize the three set approach of lifetime management.

bcarrez committed:
[SofaGeneralObjectInteraction] Update the CMakeList to initialize the three set approach of lifetime management.

bcarrez committed:
[SofaGeneralRigid] Update the CMakeList to initialize the three set approach of lifetime management.

bcarrez committed:
[SofaGeneralSimpleFem] Update the CMakeList to initialize the three set approach of lifetime management.

bcarrez committed:
[SofaGeneralTopology] Update the CMakeList to initialize the three set approach of lifetime management.

bcarrez committed:
[SofaGeneralVisual] Update the CMakeList to initialize the three set approach of lifetime management.

bcarrez committed:
[SofaGraphComponent] Update the CMakeList to initialize the three set approach of lifetime management.

bcarrez committed:
[SofaHaptics] Update the CMakeList to initialize the three set approach of lifetime management.

damienmarchal committed:
[all] Install the deprecation/stabilisation system in CMakelists

damienmarchal committed:
SofaKernel] Remove SOFA_DECL_CLASS/SOFA_LINK_CLASS

No one knows what this do.
And if we mute them this breaks nothing.
So let's deprecated that.

damienmarchal committed:
[all] Remove too general #includes in .h or .cpp and report the changes so that everything compile.

damienmarchal committed:
[SofaOpenglVisual & SofaNonUniformFem] Use the three {sets} CMakeLists approach.

damienmarchal committed:
[SofaOpenCL] It was not compiling anymore. Now it does.

damienmarchal committed:
[SceneCreator] Refactoring SceneCreator (minimal breaking)

The scene creator implementation was using the real c++ types to create
the object which was problematic in term of depency to manage to pluginize
sofa. To cut the dependencies we have implemented a new api in
sofa::simplapi:: this API helps to create scenes using the factory and strings
in a way similar to what python is doing. To make it more abstract
I also changed the SceneCreator.h signature to take BaseObject instead
of MechanicalObject.

The SceneCreator.h was also having some function we can wonder how they are related
to the creation of scene. So I move them into SceneUtils.h

Finally I propagate the changes to the files using SceneCreator.

damienmarchal committed:
[modules/sofa/component] Remove typedef/*_.h

Remove all the hardcoded typedef.
- It force inclusion of non needed elements to use them thus increasing
  code coupling and slower compilation.
- It was rarely used in the Sofa code base.

damienmarchal committed:
[tutorials] Deprecated the c++ tutorials.

These tutorials are slow to compile and very complex because they
are manipulating the raw C++ object.
So let's deprecate that and if someone have time to offer
make new ones using the sofa::simpleapi.

damienmarchal committed:
[SceneCreator] Refactor massSpringSpring to remove last dependency to sofa modules.

damienmarchal committed:
[SceneCreator] Having fun with simpleapi.

damienmarchal committed:
Merge remote-tracking branch 'origin/plugnize_sofasphfluid' into deprecating_components

# Conflicts:
#	SofaAdvanced/CMakeLists.txt
#	SofaAdvanced/config.h.in
#	SofaKernel/cmake/Modules/FindSOFA.cmake
#	applications/plugins/CMakeLists.txt
#	applications/plugins/SceneCreator/CMakeLists.txt
#	applications/plugins/SceneCreator/SceneCreator.cpp
#	applications/plugins/SofaCUDA/CMakeLists.txt
#	applications/plugins/SofaImplicitField/components/geometry/SphericalField.h
#	applications/plugins/SofaMiscCollision/CMakeLists.txt
#	applications/plugins/SofaMiscCollision/initMiscCollision.cpp
#	modules/SofaComponentAdvanced/CMakeLists.txt
#	modules/SofaComponentAdvanced/initComponentAdvanced.cpp
#	modules/sofa/component/typedef/ForceField_double.h
#	modules/sofa/component/typedef/ForceField_float.h
#	modules/sofa/component/typedef/Mapping_combination.h
#	modules/sofa/component/typedef/Mapping_double.h
#	modules/sofa/component/typedef/Mapping_float.h
#	modules/sofa/component/typedef/ProjectiveConstraintSet_double.h
#	modules/sofa/component/typedef/ProjectiveConstraintSet_float.h
#	modules/sofa/component/typedef/Topology_combination.h
#	modules/sofa/component/typedef/Topology_double.h
#	modules/sofa/component/typedef/Topology_float.h
#	modules/sofa/component/typedef/_Miscellaneous_double.h
#	modules/sofa/component/typedef/_Miscellaneous_float.h

damienmarchal committed:
[SofaMiscCollision] Fix CMakeLists.

damienmarchal committed:
Merge remote-tracking branch 'origin/fixImplicitSurfaceContainer' into deprecating_components

# Conflicts:
#	applications/plugins/CMakeLists.txt

damienmarchal committed:
[SofaKernel] Remove the macro from make config.in so that we don't need to recompile all sofa when switching from one compilation set to an another.

damienmarchal committed:
[SofaKernel & binding] add a "dirty" way to report in the scene graph component creation failure.

I'm fully aware that this may not be a super clean way to do it but I is
better to have it to help people with deprecation of component.

damienmarchal committed:
[runSofa] Display the InfoComponent in a specific way.

damienmarchal committed:
[all] remove the SOFA_LINK_CLASS

damienmarchal committed:
[SofaKernel] Add missing InfoComponent.

This component is used to report message to the user.

damienmarchal committed:
[CMakeList] flip the order of plugin vs scenecreator.

ChristianDuriez committed:
WARNING: ParallelCGLinearSolver deprecated

bcarrez committed:
add SOFA_BUILD_DEPRECATEDCOMPONENTS and SOFA_BUILD_FULLSETCOMPONENTS C++ macros

fredroy committed:
[SofaOpenGLVisual] Set deprecated components

ChristianDuriez committed:
WARNING: Supress SMP Support (no more maintained)

Suppress all the code under
#ifdef SMP...
#endif

olivier-goury committed:
Deprecate components from SofaMiscMapping and SofaUserInteraction

fredroy committed:
[SofaKernel] remove LaparoscopicRigidType (definitely because LaparoscopicType is not used/dont compiled for a really long time

epernod committed:
Move LennardJonesFF to deprecated. Split elasticity test to move lennardJones test inside its own class test. TODO: sympletic solver were using that test, need to find another ff to test for those solvers.

epernod committed:
FIX: compilation, one include of laparoRigidType was still present.

damienmarchal committed:
externalize the deprecation list.

ErwanDouaille committed:
[Deprecation] deprecate some SofaUserInteraction components with Sebastien Nelissen

epernod committed:
Move LineBendingSprings and ForceMaskOff to deprecated

ChristianDuriez committed:
Remove SMP

ChristianDuriez committed:
Remove SMP

guparan committed:
ADD deprecated components

ErwanDouaille committed:
[Deprecation] deprecate anothers components from SofaUserInteraction with Sebastien Nelissen

Commit:
[Deprecation] deprecate another component from SofaUserInteraction

epernod committed:
Move VaccumSphereForceField and UniformVelocityDampingForceField to deprecated

fredroy committed:
[Sofa] Remove Taucs

damienmarchal committed:
[ObjectCreator] Fix a compilation because of wrong integration of checker in objectcreator.

guparan committed:
Deprecate BuoyantForceField + FIX deprecatedcomponents

Commit:
UPDATE: deprecated MechanicalStateControllerOmni in deprecatedcomponents.cpp

epernod committed:
UPDATE: remove cuttingPoint, this class does nothing. No need to keep it.

damienmarchal committed:
[SofaKernel] Remove the SOFA_BUILD_*COMPONENTS from config.h and put that into a build_option_componentset.h.in

Otherwise each time we want to change profile we need to recompile everything.
With this one we just recompile what really use the SOFA_BUILD_COMPONENTS macro.

damienmarchal committed:
[SofaConstraint] remove useless SOFA_DECL_CLASS

damienmarchal committed:
[SofaGeneralExplicit] remove useless SOFA_DECL_CLASS

guparan committed:
[SofaMiscFem] FIX missing newmat dependency

guparan committed:
[SofaCUDA] FIX FixParticlePerformer include

damienmarchal committed:
[modules] Updates CMakeLists.txt to allow compilation with core profile (WIP)

bcarrez committed:
[SOFA_BUILD_COMPONENTSET] add new 4 levels: core/light/standard/full

guparan committed:
Merge branch 'deprecating_components' into stc4_base

damienmarchal committed:
[SofaGraphComponent] Fix SceneCheckAPIChange that was not installing the default checkers by defaults.

damienmarchal committed:
[SceneCreator] Remove compilation warning because of un-used parameter.

damienmarchal committed:
[SceneCreator] Fix a typo causing compilation problem.

damienmarchal committed:
[SofaMiscTopology] Update a failing test because of pluginization of RequirePlugins

NB: We really need a tools to tell user that old components are now in plugins.

damienmarchal committed:
[CmakeLists] Adds a warning to indicate when the component set value is not valid.

damienmarchal committed:
[all] Fix some failing tests because of STC

damienmarchal committed:
[SofaKernel] Add a way to give more details when object creation is not possible.

damienmarchal committed:
[SofaKernel] Dirty split of the deprecated component and the one that just provides info

A more general system should be added.
Eg:
std::string wtf(ComponentName)

Should return:
ComponentName is in SofaThisPlugin.
But you shouldn't use it because it is now deprecated .

damienmarchal committed:
[CMakeLists] Compile sofa with the "full" component set by default

Because we want to validate as much as possible as components.

damienmarchal committed:
[CMakeLists] update the comments that were outdated.

damienmarchal committed:
[SofaBoundaryCondition] Do not deprecate UniformVelocityDampingForceField now ...it is used by compliant.

damienmarchal committed:
[SofaCUDA] Restore the old path. This branch is not yet pluginized.

damienmarchal committed:
[CMakeLists] We do not compile runSofa in "core" or "light" profile

damienmarchal committed:
[CMakeLists] Disable qt gui in core/standard.

damienmarchal committed:
[SofaKernel] Improve a bit the error message when the ObjectFactory is unable to create a component.

damienmarchal committed:
[SofaKernel] Fix a crash when the factory is not able to create an InfoComponent object.

damienmarchal committed:
[SceneCreator] Restore the SceneCreators_tests

damienmarchal committed:
[SceneCreator] Fix the tests & add test for simpleapi.

damienmarchal committed:
[SofaBoundaryCondition] Fix failling test because of change in scenecreator.

SceneCreator is now using the factory to create Object. This means that
the factory must be loaded with the needed plugins. I added a loadPlugins.cpp
file to do that.

damienmarchal committed:
[SofaCUDA] Fix compilation problem introduced by sprint.

damienmarchal committed:
[SceneCreator] Fix type on StiffSprinForceField, remove useless code and autoload the library.

damienmarchal committed:
[SofaKernel] SofaImplicitOdeSolver add initXXX() to load the factory.

Otherwise the object may not be registered and thus cannot be created.

damienmarchal committed:
[runSofa] Update the test to take into account that Sofa_test is now preloading plugin.

damienmarchal committed:
[SofaGeneralEngine] Update the code in Vertex2Frame.

There is legitimate use of the component with a zero vertex array (a startup).
So we shouln't generate an error.

damienmarchal committed:
[SofaKernel] Use the right function to initialize the plugin directory.

The old code was not working because there is difference in where the plugins are
on windows & linux.

damienmarchal committed:
[Image_test] Remove the ASSERT_NO_THROW to get an usefull message.

damienmarchal commented:
@fredroy, @guparan  In this PR just wanted to have the exception message related with the Image_test.

Here it is: 
```
unknown file
C++ exception with description "[instance(0,0,0,0,0x0,non-shared)] CImg<unsigned char>::load(): Failed to recognize format of file '/Users/sofa/jenkins/workspace/mac_clang-3.4_options/pr/all/applications/plugins/image/image_test/scenes/lena.jpg'." thrown in the test body.
```

Any idea why ? Could it be the image CImg library compiled without jpg support or something like that ? 


EDIT: Ok I fixed this one...it is because ImageMagick is not install on some machine and thus CImg does not handle jpg. @guparan (I have installed ImageMagick on sofa-centos). 

EDIT2: I updated the mac. I remember that CImg need to install external tool on windows. But don't remember which one...maybe it was imagemagick also. @guparan would you have some time to dig into that ?

damienmarchal commented:
I also would like to get rid of the failure in Utils:
https://ci.inria.fr/sofa-ci/job/centos_clang-3.4_options/1522/testReport/

Here is the associated msg_warning:
[WARNING] [Utils::widenString()] Invalid or incomplete multibyte or wide character

Could it be because this one is changing the locale and depending on the system things goes wrong with the french character used in the string ?

fredroy commented:
@damienmarchal 
According to the CMakeLists.txt, image does not enable jpeg support in CImg (it needs to add "-Dcimg_use_jpeg" compiler flag)

damienmarchal committed:
[SofaKernel] Adds a toSecond function in CTime.

damienmarchal committed:
[SofaKernel] Attemps to fix the failing FileMonitor tests.

I change the test so they wait until the initial file is created before
testing anything.

damienmarchal commented:
@fredroy thanks for the info...

Here are the corresponding lines on windows: 
Adding Plugin CImgPlugin
-- Could NOT find JPEG (missing: JPEG_LIBRARY JPEG_INCLUDE_DIR) 
-- Could NOT find TIFF (missing: TIFF_LIBRARY TIFF_INCLUDE_DIR) 

So we either need to install the imagemagick approach (I assume when CIMg cannot do something with a file it calls imagemagick) or to install the JPEG/TIFF libraries on windows/centos/macos. 

I also wonder if CMakeLists.txt use conditionnal for JPEG why the tests are done unconditionally ?

fredroy commented:
Ah, you must be careful to not confuse image and CImgPlugin. Both of them are using CImg but they are using it separately.
The warning you are noticing comes from CImgPlugin (which warns if it does not find libjpeg and libtiff).
CImgPlugin_test tests if JPEG or TIFF have been enabled (if not, of course it does not test it)

As for plugin image, I think it was developed mainly by people using imagemagick on Linux or mac (which is called automatically if CImg does not recognize the format I think).
On Windows, it is not really natural to have imagemagick into the PATH...

Anyway, the solutions are:
 - change lena.jpg by lena.bmp (it should not need any external library)
 - add imagemagick to the Windows CI
 - add libjpeg/libtiff to the external libraries of Sofa (like libpng) and add libjpeg/libpng support to the image plugin.

damienmarchal committed:
[SofaKernel] FIX the randomly FileMonitor_test failure on macos (WIP)

damienmarchal committed:
[SofaKernel] Fix failure in FileMonitor (WIP)

damienmarchal committed:
[SofaKernel] Fix Failing test on FileMonitor (WIP)

damienmarchal committed:
[SofaKernel] Fix failing FileMonitor tests (WIP)

damienmarchal committed:
[SofaKernel] Fix FileMonitor test on macos (WIP)

The possible cause is the fact that the monitor boolean was just "not initialized"
to false by default.

damienmarchal committed:
[SofaKernel] FIX FileMonitor_test failure (WIP)

damienmarchal commented on SofaKernel/framework/sofa/helper/system/FileMonitor_macosx.cpp:
```diff
@@ -119,17 +119,17 @@ These steps are explained in more detail in the sections that follow.
-    FSEventStreamRef    m_eventStream;
-    bool                m_changed;
+    FSEventStreamRef    m_eventStream ;
+    bool                m_changed {false} ;
```
But alors your were not initialized ?

damienmarchal committed:
[SofaKernel] FIX FileMonitor_test failure (DONE)

The problem on MacOS was because of a not properly initialized variable.

damienmarchal committed:
[SofaKernel] Fix FileMonitor_test failures (DONE)
